### ISSUE TITLE
Expected object of scalar type Long but got scalar type Float for sequence element 1 in sequence argument at position #1 'tensors'

### ISSUE THREAD

**Issue by nikogamulin** [aladdinpersson/Machine-Learning-Collection#18]
Hi,

I rewrote the code along with watching your tutorial. When I run the training procedure, I get the following error:

```
Traceback (most recent call last):
  File "/home/niko/programs/pycharm-community-2019.2.1/helpers/pydev/pydevd.py", line 1415, in _exec
    pydev_imports.execfile(file, globals, locals)  # execute the script
  File "/home/niko/programs/pycharm-community-2019.2.1/helpers/pydev/_pydev_imps/_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "/home/niko/workspace/pytorch-and-lightning-tutorials/yolo/train_original.py", line 147, in <module>
    main()
  File "/home/niko/workspace/pytorch-and-lightning-tutorials/yolo/train_original.py", line 126, in main
    train_loader, model, iou_threshold=0.5, threshold=0.4
  File "/home/niko/workspace/pytorch-and-lightning-tutorials/yolo/utils.py", line 255, in get_bboxes
    true_bboxes = cellboxes_to_boxes(labels)
  File "/home/niko/workspace/pytorch-and-lightning-tutorials/yolo/utils.py", line 322, in cellboxes_to_boxes
    converted_pred = convert_cellboxes(out).reshape(out.shape[0], S * S, -1)
  File "/home/niko/workspace/pytorch-and-lightning-tutorials/yolo/utils.py", line 315, in convert_cellboxes
    (predicted_class, best_confidence, converted_bboxes), dim=-1
RuntimeError: Expected object of scalar type Long but got scalar type Float for sequence element 1 in sequence argument at position #1 'tensors'
```

Then I tried to copy the exact same code from your train.py and dataset.py file but the error still persisted. I guess __getitem__ in dataset.py should return long instead of float types for bounding boxes. Do you know what might be the cause of the error above?

**Comment by nikogamulin** [comment]
Are you sure that the code ran successfully? Which version of PyTorch are you running?

in convert_cellboxes, an error occurs while trying to concatenate tensors of different types:


```
predicted_class = predictions[..., :20].argmax(-1).unsqueeze(-1)
    best_confidence = torch.max(predictions[..., 20], predictions[..., 25]).unsqueeze(
        -1
    )
    converted_preds = torch.cat(
        (predicted_class, best_confidence, converted_bboxes), dim=-1
    )

```

predicted class is long, whereas best_confidence and converted_bboxes are float.

In order to run the code successfully, I had to convert predicted_class to float:

```
converted_preds = torch.cat(
        (predicted_class.float(), best_confidence, converted_bboxes), dim=-1
    )
```

Also, there was an error while calculating precision and recall for the same reason (dividing long). So I had to modify the code as follows:

```
precisions = torch.div(TP_cumsum, (TP_cumsum + FP_cumsum + epsilon))
precisions = torch.cat((torch.tensor([1.0]), precisions))
recalls = torch.cat((torch.tensor([0.0]), recalls))
```

one strange thing that I noticed was that torch.divide was not recognized, so I had to use torch.div which according to PyTorch is an alias.

Otherwise wonderful tutorials: I would appreciate if you could tell which version of PyTorch you used as although the modified code seems to be working, I would like to know how it was possible to run the original version from your repo.

**Comment by aladdinpersson** [comment]
Hey Niko, the code runs for me on PyTorch preview nightly, specifically 1.8.0.dev. However when coding this I think that was 1.7.0.dev so it should probably be able to run on 1.7.0? I think I need to figure out a way to make sure code is at least runnable on different versions of pytorch but not sure how to set this up yet. I do believe it makes sense to ensure they have the same types though, like you are doing, but when you don't encounter the problem you don't fix it 😅

**Comment by nikogamulin** [comment]
Thank you, Aladdin! I installed version 1.7.0 using pip but don't know why the running version didn't match the one from the pip list. When I checked the version while executing (print(torch.__version__)), it printed 1.4.0. Then I created conda environment and there the running version matched the installed one (1.7.0) and everything worked fine without modifications. Thank you nevertheless for quick response!
